### ISSUE TITLE
Adopt idiomatic Go repository layout

### ISSUE THREAD

**Issue by stapelberg** [restic/restic#1116]
(This is coming out of issue #1085)

Currently, restic uses `gb`, which makes it super-awkward to use with `go get` and tools which assume the `go get` layout, notably `dh-golang` for Debian packaging.

I recommend doing the following changes:

1. Move `src/cmds/restic` to `cmd/restic`
2. Move `src/restic/*` to `internal/`
3. Update all imports to use `github.com/restic/restic/internal/*`

The `internal` path component results in other packages not being able to accidentally depend on your code, which I assume is your intention, as you don’t link to godoc :).

This change can be done in an automated fashion with the following commands:

```
% mkdir cmd internal
% mv src/cmds/restic cmd/
% mv src/restic/* internal/
% sed -i 's,^\s*"restic\([^"]*\)"$,"github.com/restic/restic/internal\1",g' $(pcregrep -l -r '^\s*"restic[^"]*"$' cmd internal | tr '\n' ' ')
% sed -i 's,"restic/test","github.com/restic/restic/internal/test",g' $(pcregrep -l -r '^\s*(?:[a-z.]+ )?"restic[^"]*"$' cmd internal  | tr '\n' ' ')
% sed -i 's,^import "restic\([^"]*\)"$,import "github.com/restic/restic/internal\1",g' $(pcregrep -l -r '^import "restic[^"]*"$' cmd internal | tr '\n' ' ')
% sed -i 's,"restic/fs","github.com/restic/restic/internal/fs",g' cmd/restic/cmd_mount.go
```

Afterwards, users will be able to use `go get -u github.com/restic/restic/cmd/restic` to install/update restic.

While at it, I recommend switching to `dep` for vendoring. Maybe https://github.com/gophercon/2017-talks/tree/master/lightningtalks/CarolynVanSlyck-DepIn10 is helpful for that. Let me know if you need any help with it.

**Comment by fd0** [comment]
Yeah, good proposal. At the time `gb` appeared very convincing, and I liked the single-project workspace approach. It's a bit more, the CI tests need to be changed slightly, and the `build.go` needs to be adapted. Shouldn't be too much work. The open PRs need to be changed also. I'd also like to switch to `dep`.

So, I agree. I'm not sure what the timeline is though.

**Comment by stapelberg** [comment]
Glad you agree. The sooner this gets done, the easier it’ll be to work on the Debian packaging :).

**Comment by zcalusic** [comment]
I'm also very glad this is going in the right direction. Ever since I've opened #660 Should motivate me to work more on restic in the future. :)

**Comment by fd0** [comment]
Wow. I've just moved everything around in a branch, removed the `vendor` folder, ran `dep init`, and that's it! Very smooth! I'm **impressed**!

**Comment by stapelberg** [comment]
Yes, @sdboyer et al. have done good work on `dep` :)

**Comment by middelink** [comment]
Yeah, the only thing we have lost is the ability to compile restic from our github forks without using build.go, as the tree now refers hardcoded to github.com/restic/restic . But then again, that didn't really work in the first place either, so not a large loss :P

BTW, for some reason the master branch results in a 8722016 byte restic, but the switch-to-default-go-git-layout gives a 8861984 byte binary? Looks like you did not lock in the exact same versions of the vendored pkgs as you had in the master branch?

**Comment by fd0** [comment]
> Yeah, the only thing we have lost is the ability to compile restic from our github forks without using build.go, as the tree now refers hardcoded to github.com/restic/restic . But then again, that didn't really work in the first place either, so not a large loss :P

The standard workaround is to check out the forked repo at `github.com/restic/restic` in your GOPATH. That's an annoyance with all Go projects unfortunately.

> BTW, for some reason the master branch results in a 8722016 byte restic, but the switch-to-default-go-git-layout gives a 8861984 byte binary? Looks like you did not lock in the exact same versions of the vendored pkgs as you had in the master branch?

No, I've just used `dep` which may have changed the revisions slightly. So a different binary is expected.

**Comment by fd0** [comment]
This is merged now. One annoying thing is that the package `restic` which contains many top-level types now has an import path of `github.com/restic/restic/internal` (the last part of the path does not correspond to the package name). Any idea on how to solve this? Two solutions come to my mind:
 * Move `internal/*.go` to `internal/restic/`
 * Move `internal/*` to `internal/restic`

In both cases the path becomes `github.com/restic/restic/internal/restic`, or `github.com/restic/restic/internal/restic/backend[...]` for other packages.

Further ideas?

**Comment by zcalusic** [comment]
My vote for the first solution (i.e. moving only toplevel go files). if I understand correctly, leaving the other subfolders where they are now would keep their import paths shorter.

**Comment by tmcarr** [comment]
Great to hear dep is coming!!!!

**Comment by stapelberg** [comment]
> In both cases the path becomes github.com/restic/restic/internal/restic, or github.com/restic/restic/internal/restic/backend[...] for other packages.

Why is that an issue? I mean, sure, it’s a lot of repetition of “restic”, but `goimports` takes care of adding the import, so you never have to type it out.

I don’t really care much either way. This seems like such a little detail :).

**Comment by fd0** [comment]
Merged as #1129
